internal/store/device.go

```diff
@@ -301,64 +303,62 @@ func (s *DeviceStore) UpdateAnnotations(ctx context.Context, orgId uuid.UUID, na
-	var deviceOwner string
-	var templateVersion *model.TemplateVersion
-	err := s.db.Transaction(func(innerTx *gorm.DB) (err error) {
-		device := &model.Device{Resource: model.Resource{OrgID: orgId, Name: name}}
-		result := innerTx.First(device)
+func (s *DeviceStore) UpdateRendered(ctx context.Context, orgId uuid.UUID, name string, rendered string) error {
+	return s.db.Transaction(func(innerTx *gorm.DB) (err error) {
+		existingRecord := model.Device{Resource: model.Resource{OrgID: orgId, Name: name}}
+		result := innerTx.First(&existingRecord)
 		if result.Error != nil {
 			return flterrors.ErrorFromGormError(result.Error)
 		}
+		existingAnnotations := util.LabelArrayToMap(existingRecord.Annotations)
 
-		if device.Owner == nil {
-			return flterrors.ErrResourceOwnerIsNil
-		}
-
-		currentTemplateVersion := ""
-		annotations := util.LabelArrayToMap(device.Annotations)
-		v, ok := annotations[model.DeviceAnnotationTemplateVersion]
+		var currentRenderedVersion int64 = 0
+		renderedVersionString, ok := existingAnnotations[model.DeviceAnnotationRenderedVersion]
 		if ok {
-			currentTemplateVersion = v
+			currentRenderedVersion, err = strconv.ParseInt(renderedVersionString, 10, 64)
+			if err != nil {
+				return err
+			}
 		}
 
-		if currentTemplateVersion == "" {
-			return flterrors.ErrTemplateVersionIsNil
-		}
+		currentRenderedVersion++
+		existingAnnotations[model.DeviceAnnotationRenderedVersion] = strconv.FormatInt(currentRenderedVersion, 10)
+		annotationsArray := util.LabelMapToArray(&existingAnnotations)
 
-		if knownOwner != nil && knownTemplateVersion != nil && *device.Owner == *knownOwner && currentTemplateVersion == *knownTemplateVersion {
-			return nil
-		}
+		result = innerTx.Model(existingRecord).Updates(map[string]interface{}{
+			"annotations":     pq.StringArray(annotationsArray),
+			"rendered_config": &rendered,
+		})
 
-		templateVersion = &model.TemplateVersion{
-			ResourceWithPrimaryKeyOwner: model.ResourceWithPrimaryKeyOwner{OrgID: orgId, Owner: device.Owner, Name: currentTemplateVersion},
-		}
-		result = s.db.First(templateVersion)
-		if result.Error != nil {
-			return flterrors.ErrorFromGormError(result.Error)
-		}
-		return nil
+		return flterrors.ErrorFromGormError(result.Error)
 	})
+}
 
-	if err != nil {
-		return nil, err
+func (s *DeviceStore) GetRendered(ctx context.Context, orgId uuid.UUID, name string, knownRenderedVersion *string) (*api.RenderedDeviceSpec, error) {
+	device := model.Device{
+		Resource: model.Resource{OrgID: orgId, Name: name},
+	}
+	result := s.db.First(&device)
+	if result.Error != nil {
+		return nil, flterrors.ErrorFromGormError(result.Error)
 	}
 
-	if templateVersion == nil {
-		return nil, nil
+	annotations := util.LabelArrayToMap(device.Annotations)
+	renderedVersion, ok := annotations[model.DeviceAnnotationRenderedVersion]
+	if !ok {
+		return nil, flterrors.ErrNoRenderedVersion
```

Comment:
question: should we wrap this error with some of the context of the call to help debug? Maybe the name and or orgId.

Reply:
Until now I return the raw error from this layer and then add context in the caller